importccl: Support `IMPORT MYSQLDUMP file`

dt opened the pull request:
This adds a new `IMPORT` syntax:
```
IMPORT MYSQLDUMP (s3://path/to/some/file.sql) 
```
as well as variant to select a single table:
```
IMPORT TABLE simple FROM MYSQLDUMP (s3://path/to/some/file.sql) 
```

This parses the dump file to read the table schema(s), converting their the mysql-dialect `CREATE TABLE` statements to cockroach's corresponding table definitions.

When a tables named, only the named table is imported -- if the dump contains other tables, they are ignored.

The supported conversions of the various MySQL column types to corresponding Cockroach column types is somewhat documented by Cockroach `CREATE` statement in the `everything`  test case of `TestMysqldumpSchemaReader`, which is tested to correspond to the example in the `everything.sql` testdata file, though these mappings remain subject to change as we test the behaviors of the corresponding types.

This is a big step towards our mysqldump support, though we will be continuing to actively work on this for the near future, so this is a rough, initial version that we can continue to polish. Future work should allow reading create table definitions _during_ sampling, rather than prior to starting the DistSQL job, to eliminate the duplicate read cost.

dt committed:
importccl: convert MySQL CREATE TABLE to cockroach

This adds utilities for converting MySQL's `CREATE TABLE` statements to
Cockroach's TableDescriptors with corresponding columns and indexes.

The bulk of this is mapping MySQL's column types to cockroach's, which
is not quite 1:1 particularly with regard to length/width specifiers.

As a first approximation, numeric types are mapped to their nearest
counterparts (ints to int, floats to floats, decimal to decimal) using
what appear to be the closest size-specifying aliases (e.g. smallint),
with scale and precision arguments converted as well. For more detailed
mapping of input-to-converted types, compare the schemas in the files in
`testdata`: The `everything` test  attemps to verify the behavior of all
supported column types -- a MySQL table containing the various columns
with the various combinations of supported options (length, width,
not/null, etc) is dumped in the `everything.sql` testdata file. When
converted to via this import process, the corresponding Cockroach schema
is represented in `everything.cockroach-schema.sql`.

Note however that the behavior of these types, or their parameters, may
not actually match between Cockroach and MySQL. There are cases where
the MySQL modifier has no cockroach counterpart, like `ZEROFILL` or
others where the behavior may differ, e.g. size handling. The exact ways
these differ as far as user-visible behavior will need to be determined
via side-by-side testing, of the mysql and imported cockroach schemas,
as part of the larger process of producing migration docs.

Release note: none.

dt committed:
sql: export InitTableDescriptor

Release note: none.

dt committed:
importccl: add detail to mysql errors

Release note: none.

dt committed:
importccl: refactor mysql testdata creation

Make it easier to support single and multi table dumps.

Release note: none.

dt committed:
importccl: 'IMPORT TABLE foo FROM MYSQLDUMP bar'

This adds support for loading a table, including its schema, by just
pointing at a mysqldump file. The IMPORT process extracts the schema
from the file, converts the MySQL types to corresponding cockroach
types, and then loads the data for that table from the dump file.

Release note (sql change): prototype support for IMPORT ... MYSQLDUMP.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26164)
<!-- Reviewable:end -->

mjibson commented on pkg/ccl/importccl/read_import_mysql.go:
```diff
@@ -143,5 +150,197 @@ func mysqlToCockroach(
+		def.Type = strOpt(coltypes.VarChar, length)
+	case mysqltypes.Text:
+		def.Type = strOpt(coltypes.Text, length)
+	case mysqltypes.Blob:
```
Thoughts about using `case mysqltypes.Blob, mysqltypes.VarBinary, ...:` to simplify this down to a single case? Don't care either way.

dt replied:
I started with that (with all the duplicate Int types grouped), but switched to one-line-per-case to make it easier to eyeball that all the original cases were being matched (can copy/paste or side-by-side with the enum def this way).

mjibson commented on pkg/ccl/importccl/read_import_mysql.go:
```diff
@@ -143,5 +150,197 @@ func mysqlToCockroach(
+	case mysqltypes.Uint32:
+		def.Type = coltypes.Int
+	case mysqltypes.Int64:
+		def.Type = coltypes.BigInt
```
Why are we using bigint here and not int4 or int2 above? Is it ok to just use int everywhere? This is hard.

mjibson replied:
My worry about attempting to preserve these as close as possible to MySQL is that they may imply that the behavior will be the same as it was. Whereas if we change all the types to be the most canonical form then I think that would encourage old MySQL users to read the documentation on our types. My opinion is that the type aliases are primarily for Postgres ORMs. I'm happy to hear another argument, but if it were me going from mysql to cockroach, I would be slightly miffed that although the displayed type names were the same, their behavior was different. I would expect the type name to change if the behavior from mysql was changing.

mjibson commented on pkg/ccl/importccl/read_import_mysql.go:
```diff
@@ -143,5 +150,197 @@ func mysqlToCockroach(
+		def.Type = coltypes.BigInt
+
+	case mysqltypes.Float32:
+		def.Type = coltypes.Float4
```
Is there a difference here? I thought floats are all 64 bit in cockroach regardless. It seems silly to keep around a mysql-ism only in the visible type when the underlying type is really just a float.

dt replied:
We have a bunch of code around preserving the different visible types for the same underlying type, so I assumed we wanted to get as close as we could?

benesch replied:
Yes, this is very important. E.g., JS doesn’t support 64-bit ints so any int8s have to be returned as strings. I’m sure there are similar problems in other languages with other drivers. Best to take the type at face value where possible. Plus switching to a wider type in Cockroach is only a metadata change so easy for the user to do later.

mjibson replied:
@benesch I'm not sure exactly what your proposal is here. Are you arguing to keep the types as appearing as-close-as-possible to mysql or to make the visible type match the underlying type?

mjibson commented on pkg/ccl/importccl/read_import_mysql.go:
```diff
@@ -143,5 +150,197 @@ func mysqlToCockroach(
+	case mysqltypes.Geometry:
+		fallthrough
+	case mysqltypes.TypeJSON:
+		fallthrough
```
we can't do this?

dt replied:
Figured punt it to a followup, since I don't know yet if/how our json type matches up.

dt replied:
added TODOs for this and `bit`.

mjibson commented on pkg/ccl/importccl/read_import_mysql.go:
```diff
@@ -143,5 +150,197 @@ func mysqlToCockroach(
+	case mysqltypes.TypeJSON:
+		fallthrough
+	case mysqltypes.Bit:
+		fallthrough
```
i thought we supported bit?

mjibson commented on pkg/ccl/importccl/testdata/mysqldump/everything-schema.sql:
```diff
@@ -0,0 +1,74 @@
+  `bl` blob,
+  `dt` datetime DEFAULT NULL,
+  `d` date DEFAULT NULL,
+  `ts` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
```
do we need to error or something because of the `on update`?

dt replied:
I think we need to doc it as one of the migration concerns, but if we're strict here and error on it (or require some option or other user intervention), I think we'll reject most schemas or make the process harder.

I'd be more inclined to do that if I thought it would be preventing accidental mistakes / bugs, but  given just how different the rest of the SQL semantics are (for example, case sensitivity), I think there's already going to be a migration process required well beyond just the data import, so I think we can lean towards being easier / a bit more lenient in the import step.

mjibson replied:
Agreed.

madelynnblue reviewed (commented):
Should we add a way for people to see what the create table statements would turn into so they can verify it'll end up correct?

mjibson commented:
Agree about the dry run. LGTM then.

dt committed:
importccl: multi-table `IMPORT MYSQLDUMP foo`

This adds support for importing all tables in a MYSQLDUMP in one IMPORT
statement, without specifying the table names or schemas separately.

When using this syntax, all table in the dump are imported.

Under the hood, the gateway node reads the whole dump to setup a multi-
table import process.  In the future though, this could be done during
the sampling phase, with parsed tables returned to the gateway with the
samples.

Release note (sql change): support IMPORT entire (multi-table) mysqldump files.

dt commented:
Okay, added multi-table IMPORT support.

mjibson commented on pkg/ccl/importccl/import_stmt_test.go:
```diff
@@ -1284,52 +1284,74 @@ func TestImportMysql(t *testing.T) {
+	const expectAll, expectSimple, expectSecond = "all", "simple", "multitable"
+
 	for _, testCase := range [][]string{
-		{`read data only`, simple, `IMPORT TABLE simple (i INT PRIMARY KEY, s text, b bytea) MYSQLDUMP DATA ($1)`},
```
i'd love to see a test with two input files. we've had errors with those before so we should make sure they don't occur again.

dt replied:
the syntax only accepts one file right now.

mjibson commented on pkg/ccl/importccl/read_import_mysql.go:
```diff
@@ -164,30 +174,40 @@ func mysqlToCockroach(
 		if i, ok := stmt.(*mysql.DDL); ok && i.Action == mysql.CreateStr {
 			name := i.NewName.Name.String()
 			if match != "" && match != name {
-				found = append(found, name)
```
does this need to stay? it's used in the error message above.

mjibson commented on pkg/ccl/importccl/read_import_mysql.go:
```diff
@@ -164,30 +174,40 @@ func mysqlToCockroach(
 			}
-			id := sqlbase.ID(int(defaultCSVTableID))
-			return mysqlTableToCockroach(evalCtx, id, name, i.TableSpec)
+			id := sqlbase.ID(int(defaultCSVTableID) + len(ret))
```
wait i forget how the table ID stuff works. IMPORT doesn't use a key rewriter. these IDs aren't valid to be inserted because they weren't allocated from a txn. so where's the thing that generates IDs for these tables that can be inserted into our kv store?

dt replied:
still in import_stmt.go, the last step before it kicks off the job: it generates ids and puts them in the descriptors.

When we switch to kicking on the job without descriptors, we'll need to do that after sampling instead and then rekey the samples.

dt commented on pkg/sql/sem/tree/name_resolution.go:
```diff
@@ -364,17 +364,28 @@ func (t *TableName) ResolveTarget(
+
+// ResolveCatalog resolves db name to a schema meta using a search path.
+func ResolveCatalog(
+	ctx context.Context, r TableNameTargetResolver, searchPath sessiondata.SearchPath, dbName string,
```
I'll let @knz chime in. Part of the difference is that `ResolveTarget` is a method _on TableName_, primarily concerned with resolving that while the arguments are things used in doing so. Here however`dbName` is the actual thing I'm interested in resolving, so it seemed odd to have it in the middle of the dependencies.

dt replied:
Done.

knz commented:
Nice work. I think it's overall all right. You don't need `ResolveCatalog`, see my comments below.

Overall I think most of the new functions would be much better if they contained numerous comments instead of none.
Disclaimer: I have not reviewed the detail of the new logic.

---

Reviewed 3 of 3 files at r5, 5 of 14 files at r7, 1 of 11 files at r8, 18 of 18 files at r9.
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/importccl/import_stmt.go, line 307 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LEPpxE58KU0ZV787iX5:-LEPpxE66DuEuHHC75Hp:bsgm3ei) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/ccl/importccl/import_stmt.go#L307)):*
> ```Go
> 			var descI tree.SchemaMeta
> 			if table == nil {
> 				found, _, descI, err = tree.ResolveCatalog(ctx,
> ```

No need for this `ResolveCatalog` stuff.

Use either `sql.ResolveDatabase(ctx, p, dbName, true /*required*/)` (recommended, uses db cache)

or `p.PhysicalSchemaAccessor().GetDatabaseDesc(dbName, p.CommonLookupFlags(ctx, true /*required*/))` (skips db cache)

---

*[pkg/ccl/importccl/read_import_mysql.go, line 176 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LEPoUiN82FxhsqPhTL4:-LEPoUiN82FxhsqPhTL5:byh1xvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/ccl/importccl/read_import_mysql.go#L176)):*
> ```Go
> 
> func readMysqlCreateTable(
> 	input io.Reader, evalCtx *tree.EvalContext, match string,
> ```

I think this function deserves comments, both at the top and through the logic.

I had the same pause as matt wrt descriptors. This needs to be explained for future readers too.

---

*[pkg/ccl/importccl/read_import_mysql.go, line 215 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LEPobR82sPrRfDwOMbv:-LEPobR82sPrRfDwOMbw:b-qtabzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/ccl/importccl/read_import_mysql.go#L215)):*
> ```Go
> }
> 
> func mysqlTableToCockroach(
> ```

ditto docs/comments

---

*[pkg/ccl/importccl/read_import_mysql.go, line 262 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LEPooW02xR5u7QznTtP:-LEPooW02xR5u7QznTtQ:b-qtabzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/ccl/importccl/read_import_mysql.go#L262)):*
> ```Go
> }
> 
> func mysqlColToCockroach(name string, col mysql.ColumnType) (*tree.ColumnTableDef, error) {
> ```

ditto docs/comments

---

*[pkg/ccl/importccl/read_import_mysql.go, line 344 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LEPox4OCYXv8U3uo186:-LEPox4OCYXv8U3uo187:b83donj) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/ccl/importccl/read_import_mysql.go#L344)):*
> ```Go
> 		fallthrough
> 	case mysqltypes.TypeJSON:
> 		// TODO(dt): is our type close enough to use here?
> ```

probably yes

---

*[pkg/ccl/importccl/read_import_mysql.go, line 347 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LEPoyLm8-7P9xv8ou2A:-LEPoyLm8-7P9xv8ou2B:b83dgqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/ccl/importccl/read_import_mysql.go#L347)):*
> ```Go
> 		fallthrough
> 	case mysqltypes.Bit:
> 		// TODO(dt): is our type close enough to use here?
> ```

probably not

---

*[pkg/ccl/importccl/read_import_mysql_test.go, line 140 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LEPp21R6-KDrB29WwT2:-LEPp21R6-KDrB29WwT3:b-8jo20q) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/ccl/importccl/read_import_mysql_test.go#L140)):*
> ```Go
> 		expected := descForTable(t, `
> 			CREATE TABLE everything (
> 				i			INT PRIMARY KEY,
> ```

nit: the column types are funkily spaced.

---

*[pkg/sql/sem/tree/name_resolution.go, line 377 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LELrY-jcKOVc5zEJ29k-r9-377:-LEPph3sDAe5-0f1Wyap:br9ynvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/sql/sem/tree/name_resolution.go#L377)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I'll let @knz chime in. Part of the difference is that `ResolveTarget` is a method _on TableName_, primarily concerned with resolving that while the arguments are things used in doing so. Here however`dbName` is the actual thing I'm interested in resolving, so it seemed odd to have it in the middle of the dependencies.
</blockquote></details>

You do not need this method *at all* to do what you need to do. See my other comment.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-:-LEPqeQj1pdQgK9H7GFW:b-v9qs6b)*
<!-- Sent from Reviewable.io -->

mjibson commented on pkg/sql/parser/sql.y:
```diff
@@ -1644,7 +1644,11 @@ import_data_format:
 // %SeeAlso: CREATE TABLE
 import_stmt:
- IMPORT TABLE table_name FROM import_bundle_format '(' string_or_placeholder ')'
+ IMPORT import_bundle_format '(' string_or_placeholder ')'
```
I think I've changed my mind on this. It's fine as is.

mjibson commented on pkg/ccl/importccl/read_import_mysql.go:
```diff
@@ -139,5 +170,215 @@ func mysqlToCockroach(
+
+	time := hlc.Timestamp{WallTime: evalCtx.GetStmtTimestamp().UnixNano()}
+	priv := sqlbase.NewDefaultPrivilegeDescriptor()
+	desc := sql.InitTableDescriptor(id, defaultCSVParentID, name, time, priv)
```
we already have a parentID up in the import function. i think we should plumb it down to here so that we can change it in one place.

dt replied:
Done.

dt commented on pkg/sql/distsql_plan_csv.go:
```diff
@@ -236,21 +245,30 @@ func LoadCSV(
-	splits = append(splits, tableSpan.Key)
+	if tableDesc != nil {
+		tableSpan := tableDesc.TableSpan()
+		splits = append(splits, tableSpan.EndKey)
```
Done.

dt commented on pkg/sql/distsql_plan_csv.go:
```diff
@@ -236,21 +245,30 @@ func LoadCSV(
+	for _, desc := range tables {
+		tableSpan := desc.TableSpan()
+		splits = append(splits, tableSpan.EndKey)
+		splits = append(splits, tableSpan.Key)
```
Done.

dt commented on pkg/sql/distsql_plan_csv.go:
```diff
@@ -195,31 +197,38 @@ func LoadCSV(
-			})
+			}
+			if tableDesc != nil {
+				spec.TableDesc = *tableDesc
```
not quite sure I follow... right now it's sometimes nil (for multi-table import) so we can't just blindly re-reference it.

mjibson commented:
ah. great. ok then.

---

Review status: 0 of 39 files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_plan_csv.go, line 220 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LEFcxzycHnFI7jEbg-N-r9-220:-LETAyIV7zDfD2Wk4tCB:b-wcgkc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/sql/distsql_plan_csv.go#L220)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

not quite sure I follow... right now it's sometimes nil (for multi-table import) so we can't just blindly re-reference it.
</blockquote></details>

You're right. I missed the `*`.

---

*[pkg/sql/distsqlrun/processors.proto, line 562 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LELrY-jcKOVc5zEJ29j-r9-562:-LETAWJX9E9oAhL14Men:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/sql/distsqlrun/processors.proto#L562)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Eh, do-able, but not trivial. I'd rather not refactor the single-table procs more in this change for now?
</blockquote></details>

:thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-:-LETA_wp5U9AaH3fKA5z:bm5474e)*
<!-- Sent from Reviewable.io -->

knz commented:
So the code  LGTM but I am concerned by the commit message and the user story here.

The commit message says "The bulk of this is mapping MySQL's column types to cockroach's, which is not quite 1:1 [...]". This is doing a disservice to both your work (there was a lot of thinking going into this already and the commit message undersells that work) and the users -- there is nothing in there to tell me:

- what is the actual mapping (I want to see a table in the commit message: one side with mysql and one side with cockroachdb)
- what entries in the mapping are dubious (things that gave you pause, or where you believe there will be special values that can cause problems)
- where the precision/width mapping is probably wrong/incomplete (I think anything other than DECIMAL has problems)
- what are the column qualifiers in mysql that have no equivalent in CockroachDB
- a clear argument of why all possible table/index/column identifiers in MySQL are also valid in CockroachDB (how do the identifier rules differ, if at all?)

Also the release note cannot be "none" for a change that has so much impact.

---

Reviewed 1 of 34 files at r10, 3 of 3 files at r11, 6 of 22 files at r13, 16 of 16 files at r15.
Review status: all files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/importccl/read_import_mysql_test.go, line 140 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LEPp21R6-KDrB29WwT2:-LETklbeDYlBjvZ6ezkO:b-tw51u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/ccl/importccl/read_import_mysql_test.go#L140)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

still shows irregularly spaced in reviewable for me.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-:-LETlAxl2Slr62PA92bt:b-mkyvxu)*
<!-- Sent from Reviewable.io -->

dt commented:
@knz I believe there are release notes on the commits that actually changed the user-visible behaviors -- is there one that is missing?

I think you're looking at https://github.com/cockroachdb/cockroach/pull/26164/commits/74e0e281bc8899ab4e549885999dda079d5ed676, which is just the helper and a unit test, without a user-facing call.

As far as user-facing usage docs, particularly w.r.t many of these type behavior difference questions, the short answer is we don't know yet: we need to run these side-by-sid with a schema in MySQL, imported it int cockroach, and then compare what it does. That process will likely revise some of the conversions,  but will be part of the larger effort to create guidance for migration from MySQL, covering both the initial IMPORT but also the most likely to be encountered semantic and behavioral differences that will follow (e.g. case (in)sensitive comparisons, unicode handling, (lack of) triggers, etc).

dt commented:
Review status: 3 of 42 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/importccl/read_import_mysql_test.go, line 140 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LEPp21R6-KDrB29WwT2:-LEUaRe3-mD4kCZVY4Bv:bhb1534) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/ccl/importccl/read_import_mysql_test.go#L140)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

still shows irregularly spaced in reviewable for me.
</blockquote></details>

Moved to stand-alone file (that is side-by-side with its mysql counterpart) and switched the file to use spaces since tabs were being a pain.

---

*[pkg/sql/distsql_plan_csv.go, line 220 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LEFcxzycHnFI7jEbg-N-r9-220:-LEUaLMI-pLpkXu9pybo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/sql/distsql_plan_csv.go#L220)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

You're right. I missed the `*`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26164)*
<!-- Sent from Reviewable.io -->

knz commented:
>  I believe there are release notes on the commits that actually changed the user-visible behaviors -- is there one that is missing?

There are currently 2 commits with a release note in the patch set. Both release notes are one-liners and do not give a sense of how much of mysql is supported. See my other point below.

> As far as user-facing usage docs, particularly w.r.t many of these type behavior difference questions, the short answer is we don't know yet

Then how have you come up with the `case` s in the giant `switch` to map the schema info from mysql to crdb? I assume this was not random work. You must have applied some reasoning / principles. My point is that this reasoning should be summarized in the commit message. 


The code LGTM further.

---

Reviewed 2 of 34 files at r16, 10 of 18 files at r17, 3 of 12 files at r18, 16 of 16 files at r19.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/ccl/importccl/testdata/mysqldump/simple.cockroach-schema.sql, line 5 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LEk3rnp0bHDeaEdEBRw:-LEk3rnq90jLGVk3Ye4v:bdr0dub) ([raw file](https://github.com/cockroachdb/cockroach/blob/193ff39375b6d78b9f709414832c98ad3cacd549/pkg/ccl/importccl/testdata/mysqldump/simple.cockroach-schema.sql#L5)):*
> ```SQLPL
>   s text,
>   b bytea
> )
> ```

nit: newline missing at end of file

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-:-LEk49dtDXgKxfj18C12:b-z3cb88)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/parser/sql.y, line 1652 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LEk6j4t8j0x6B3sy2QQ:-LEk6j4u58ItKTGraStC:b-3s0lla) ([raw file](https://github.com/cockroachdb/cockroach/blob/193ff39375b6d78b9f709414832c98ad3cacd549/pkg/sql/parser/sql.y#L1652)):*
> ```Yacc
> // %SeeAlso: CREATE TABLE
> import_stmt:
>  IMPORT import_bundle_format '(' string_or_placeholder ')'
> ```

Sorry I had forgotten to check this: both here and below an EXPERIMENTAL keyword is missing.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26164)*
<!-- Sent from Reviewable.io -->

dt commented:
@knz this is being shipped in 2.1 as a supported feature (maybe "beta"), so no experimental labels.

knz commented:
> @knz this is being shipped in 2.1 as a supported feature (maybe "beta"), so no experimental labels.

Well then doubly so a reason to ensure that the compatibility story is straight and tested ahead of time.  I don't think it does us good to ship a "beta" / non-experimental feature where we have no idea yet whether we'll need to make major changes later because our understanding of mysql schema semantics is flawed.

I would like to call a process flaw here: I don't think it is reasonable to make *any* feature that aims at compatibility non-experimental unless/until we have built good confidence that we know what works and what doesn't, what are the edge cases and what users need to care about. The feature should be clearly marked as experimental until this investigation is completed. @awoods187 thoughts?

dt commented:
@knz we have until September to mark it experimental though, *if* our testing/QA/etc suggests that that is necessary.

knz commented:
> we have until September to mark it experimental though, if our testing/QA/etc suggests that that is necessary.

1) we've never done that before
2) we don't have a process to check, when release time comes, which new features have failed to gain enough confidence to dowgrade from non-experimental to experimental
3) it seems unlikely, given the engineering culture "keep it if it's not broken" that we'll do this properly in that direction

Ultimately I am not the right person to call the shots here but I am concerned by the overall process.

dt commented:
We have definitely done that before -- we write a feature, then as we approach the release, we've evaluating its stability, compatibility, known unknowns, etc and that based on that, determined it warranted the experimental label. e.g. 15a809511

dt commented:
I fleshed out that commit message a little bit more - basically just saying that as a first pass, we just picked the closest looking types, i.e. ints are ints, floats are floats, etc and tried to preserve the params as closely as possible. This seems like a good starting point on the path to the behavior of the imported data being as similar as possible, which we'll work on via side-by-side testing.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/distsqlrun/processors.proto, line 562 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LELrY-jcKOVc5zEJ29j-r9-562:-LEkTB6nDK47UAvbgqcJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/sql/distsqlrun/processors.proto#L562)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

:thumbsup: 
</blockquote></details>

Done.

---

*[pkg/sql/parser/sql.y, line 1647 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-LEFJ4ktGDmCPCPZ_Vz5-r9-1647:-LEkTEISBaayueuPAoFa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a11ca9d95940eb9efb4e2ff94a6a0213ece059/pkg/sql/parser/sql.y#L1647)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I think I've changed my mind on this. It's fine as is.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26164#-:-LEkTFgbBm9LwaXgo0sF:b-p8ki62)*
<!-- Sent from Reviewable.io -->

dt commented:
@knz since matt and I are trying to collaborate closely on this feature, so we're trying to break it up into smaller, functional pieces that are merged early and often, as we go. One bit of context that might not be shared here is that that this PR is just _one step_ in the larger feature we're developing and planning to ship, not _the_ finished product. This feature is the main thing matt and I are scheduled to work on for the next month, so there will be lots of follow-up work, including things like testing and determining what the exact mappings between mysql and cockroach types should be, and then documenting that. Our priority right now though is to get the rough functionality in place, even if we then come back and e.g. change how we handle column types or lengths in a week or two.

knz commented:
LGTM

Thanks for the context information. It could be / have been interesting to include it in the PR description and/or commit message(s).

dt commented:
@knz yeah, sorry about that: matt an I had some implicit context based on our pre-made execution plan, which was then missing for any additional reviewers joining in.

dt commented:
bors r+
